tests/test_pipenv.py

```diff
@@ -131,3 +134,56 @@ def test_activate_virtualenv_no_source(self):
         venv = Project().virtualenv_location
 
         assert command == '{0}/bin/activate'.format(venv)
+
+    @patch('pipenv.project.Project.virtualenv_location', new_callable=PropertyMock(return_value='foo'))
```

Comment:
So... this one...
The problem was: I patched `delegator.run`, and it's was used in `which_pip` and in turn in the `which` function, when project.virtualenv_location was called it would also call `delegator.run` and I would get a error in `which` because of `Mock` object and `Str` concatenation.

One thing that I would suggest (but in another PR) is to move all the helper functions from cli.py to another place (maybe `utils.py`), because click does something do whatever is in `cli.py` that makes any function present in it inaccessible to Mock (that's why I couldn't mock `which_pip` for instance in this tests).

But I saw that `project.virtualenv_location` is no longer calling the  `delegator.run` by default (only if a given env var is present), so now I can remove this mocks for this `virtualenv_location` too =)